runtime/src/main/java/org/corfudb/runtime/LogReplicationRoutingQueueListener.java

```diff
@@ -0,0 +1,332 @@
+     */
+    // TODO: Change the param type from CorfuStreamEntries to RoutingTableEntryMsg
+    protected abstract boolean processUpdatesInLogEntrySync(List<Queue.RoutingTableEntryMsg> results);
+
+    /**
+     * Invoked by the Corfu runtime when this listener is being subscribed.  This method should
+     * read all application tables which the client is interested in merging together and perform the merge.
+     * @param txnContext transaction context in which the operation must be performed
+     */
+    protected abstract void performFullSyncAndMerge(TxnContext txnContext);
+
+    /**
+     * Name with which this client was registered using the interfaces in LogReplicationLogicalGroupClient
+     * @return client name
+     */
+    protected abstract String getClientName();
+
+    /**
+     * Callback to indicate that an error or exception has occurred while streaming or that the stream is
+     * shutting down. Some exceptions can be handled by restarting the stream (TrimmedException) while
+     * some errors (SystemUnavailableError) are unrecoverable.
+     * To be implemented on the client/application
+     * @param throwable
+     */
+
+    // On error, they have to re-subscribe. In the subscribe methods, we do the processing
+    // Or, Processing the queue
+    public void onError(Throwable throwable) {
+        // Handle Trim exception
+        // 1. Create a txn. TODO: Change it to scoped txn
+        // 2. Read all entries
+        // 3. subscribe listener with the scoped txn timestamp
+
+        log.error("Exception caught during routing queue streaming processing. Re-subscribe this listener to latest timestamp. " +
+                "Error=", throwable);
+
+        if (throwable instanceof TrimmedException) {
+            corfuStore.unsubscribeListener(this);
+            try (TxnContext tx = corfuStore.txn(namespace)) {
+                List<Table.CorfuQueueRecord> records = routingQueue.entryList();
+                for (int i = 0; i < records.size(); i++) {
+                    log.debug("Entry:" + records.get(i).getRecordId());
+                    Queue.CorfuGuidMsg recordId = records.get(i).getRecordId();
+                    Queue.RoutingTableEntryMsg entry = tx.getRecord(routingQueue, recordId).getPayload();
+                    boolean callbackResult = false;
+                    if (entry.getType() == Queue.ReplicationType.FULL_SYNC) {
+                        callbackResult = processUpdatesInSnapshotSync(Collections.singletonList(entry));
+                    } else if (entry.getType() == Queue.ReplicationType.LOG_ENTRY) {
+                        callbackResult = processUpdatesInLogEntrySync(Collections.singletonList(entry));
+                    }
+                    if (callbackResult) {
+                        // When the client processed the entry successfully, delete the entry from the routing queue.
+                        tx.delete(routingQueue, recordId);
+                    }
+                }
```

Comment:
we can have this block of code in onNext() instead of listening on LR status table